### PR TITLE
feat: losses and metrics

### PR DESCRIPTION
**Pull request by hlee9212**
Hi Niels, this addresses the [the first question](https://github.com/thousandbrainsproject/tbp.tbs_sensorimotor_intelligence/pull/26#discussion_r2123690476) from #26. 

## What's Added?
Two loss functions:
- `quaternion_geodesic_loss`
- `masked_cross_entropy_loss` (This will be used for continual learning model - I'll open PR soon)

Metric:
- `get_rotation_error_in_degrees`

## Notes
1. Mathematically, the `quaternion_geodesic_loss` and `get_rotation_error_in_degrees` computes the same thing, just that the former is differentiable to torch, and the latter is closer to how Monty calculates rotation error for fair comparison. The metric also handles in case there is some `np.nan`, when I was handling really bad quaternion predictions which happened in very long epochs.

### CODE REVIEW

File: pytorch/src/losses/loss.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def masked_cross_entropy_loss(
+    logits: torch.Tensor, targets: torch.Tensor, valid_class_indices: torch.Tensor
+) -> torch.Tensor:
+    """Compute masked cross-entropy loss for continual learning.
+
+    This function calculates the cross-entropy loss only for valid classes specified by
+    valid_class_indices. It masks out invalid classes by setting their logits to a very
+    negative value before applying softmax.
```
Comment (nielsleadholm) [inline pytorch/src/losses/loss.py:83]: **suggestion (non-blocking):** You could add a comment along-the-lines of "This is a global mask, i.e. there is no granularity in the form of different masks for different members of the batch."

File: pytorch/src/metrics/rotation_error.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    pass
+
+
+def get_rotation_error_in_degrees(
+    predicted_quaternion: torch.Tensor,
+    target_quaternion: torch.Tensor,
+    reduce: bool = False,
+) -> torch.Tensor:
+    """Get the rotation error between two quaternions.
```
Comment (nielsleadholm) [inline pytorch/src/metrics/rotation_error.py:31]: **suggestion:** It would be worth explicitly documenting what quaternion format is being used in the Pytorch code / ground-truth data up until now. Does it match the SciPy expectation?

**Review by nielsleadholm — COMMENTED** [review]
Looks great, thanks for the clean code and thorough documentation! Just suggested a clarification about quaternion standardization.

Again would appreciate if @scottcanoe you can cast an eye over this one.

**Review by scottcanoe — APPROVED** [review]
Looks good! Side question: how often would you guess zero-norm quaternions make their way into `get_rotation_error_in_degrees`? Is that something that would happen more earlier (after little training)?
